packages/tlsync/src/lib/squish.ts

```diff
@@ -0,0 +1,174 @@
+import { UnknownRecord } from '@tldraw/store'
+import { exhaustiveSwitchError, objectMapEntries, structuredClone } from '@tldraw/utils'
+import {
+	NetworkDiff,
+	ObjectDiff,
+	RecordOp,
+	RecordOpType,
+	ValueOpType,
+	applyObjectDiff,
+} from './diff'
+import { TLSocketServerSentDataEvent } from './protocol'
+
+interface State<R extends UnknownRecord> {
+	lastPatch: (TLSocketServerSentDataEvent<R> & { type: 'patch' }) | null
+	squished: TLSocketServerSentDataEvent<R>[]
+}
+
+type Bailed = boolean
+
+function patchThePatch(lastPatch: ObjectDiff, newPatch: ObjectDiff): Bailed {
+	for (const [newKey, newOp] of Object.entries(newPatch)) {
+		switch (newOp[0]) {
+			case ValueOpType.Put:
+				lastPatch[newKey] = newOp
+				break
+			case ValueOpType.Append:
+				if (lastPatch[newKey] === undefined) {
+					lastPatch[newKey] = newOp
+				} else {
+					const lastOp = lastPatch[newKey]
+					if (lastOp[0] === ValueOpType.Append) {
+						const lastValues = lastOp[1]
+						const lastOffset = lastOp[2]
+						const newValues = newOp[1]
+						const newOffset = newOp[2]
+						if (newOffset === lastOffset + lastValues.length) {
+							lastValues.push(...newValues)
+						} else {
+							// something weird is going on, bail out
+							return true
+						}
+					} else {
+						// bail out, it's too hard
+						return true
+					}
```

Comment:
if the last op is a PUT it should be ok to just append the values, right?

And also if the last op was a DELETE it should be an error rather than a bail.

Reply:
good idea, added a clause to handle that

see below re: assertions